### PR TITLE
Feat: Allow raising errors from within a custom whatever(#2718)

### PR DESCRIPTION
**Pull request by abdalla-rko**
### Reasons for making this change

I introduced a **raiseFieldErrors** method to raise errors from within a custom component(see #2718). 

The current behavior allows errors to be raised from custom components and when the form **onChange** is triggered, we clear the custom raised errors. I appreciate any feedback on this approach. If everything looks good, I'll write tests and implement this in other themes.

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [ ] **I'm adding or updating code**
  - [ ] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by heath-freenome** [comment]
@abdalla-rko If the goal is to update the errorSchema via a custom widget, maybe it makes sense to update the `onChange` in `WidgetProps` to function similarly to the `onChange` in FieldProps, where the `errorSchema` can be updated as it is flows through the schema hierarchy. Feel free to join the next meeting so we can talk about it. Or reach out to me on Discord

**Comment by nickgros** [comment]
@abdalla-rko are you still working on this?

**Comment by abdalla-rko** [comment]
Hi @heath-freenome, Sorry for taking too long. I think the initial implementation is now like you suggested updating the **errorSchema** using **onChange** in the custom widget. Let me know if it's good and then I'll update the docs and other themes if needed.

File: packages/core/src/components/Form.tsx [hunk 1]
```diff
@@ -575,12 +585,35 @@ export default class Form<
...
+    const filteredErrors: ErrorSchema<T> = _pick(schemaErrors, fieldNames as unknown as string[]);
+    // Removing undefined and empty errors.
+    const filterUndefinedErrors = (errors: any): ErrorSchema<T> => {
+      Object.keys(errors).forEach((key: string) => {
+        const errorKey = key as keyof typeof errors;
+        if (errors[errorKey] === undefined) {
+          delete errors[errorKey];
+        } else if (typeof errors[errorKey] === 'object' && !Array.isArray(errors[errorKey].__errors)) {
+          filterUndefinedErrors(errors[errorKey]);
+        }
```
Comment (heath-freenome) [inline packages/core/src/components/Form.tsx:610]: Use a local variable to reduce the operations slightly
```suggestion
        const errorAtKey = errors[errorKey];
        if (errorAtKey === undefined) {
          delete errors[errorKey];
        } else if (typeof errorAtKey === 'object' && !Array.isArray(errorAtKey.__errors)) {
          filterUndefinedErrors(errorAtKey);
        }
```

File: packages/core/src/components/Form.tsx [hunk 2]
```diff
@@ -575,12 +585,35 @@ export default class Form<
   omitExtraData = (formData?: T): T | undefined => {
     const { schema, schemaUtils } = this.state;
     const retrievedSchema = schemaUtils.retrieveSchema(schema, formData);
+
     const pathSchema = schemaUtils.toPathSchema(retrievedSchema, '', formData);
```
Comment (heath-freenome) [inline packages/core/src/components/Form.tsx:589]: Revert the space
```suggestion
    const retrievedSchema = schemaUtils.retrieveSchema(schema, formData);
    const pathSchema = schemaUtils.toPathSchema(retrievedSchema, '', formData);
```

File: packages/utils/src/types.ts [hunk 3]
```diff
@@ -750,7 +750,7 @@ export interface WidgetProps<T = any, S extends StrictRJSFSchema = RJSFSchema, F
   /** The input blur event handler; call it with the widget id and value */
   onBlur: (id: string, value: any) => void;
   /** The value change event handler; call it with the new value every time it changes */
-  onChange: (value: any) => void;
+  onChange: (value: any, es?: ErrorSchema<T>, id?: string) => void;
```
Comment (heath-freenome) [inline packages/utils/src/types.ts:753]: I don't see any code that grabs these errors from the widget and passes it up the chain. If you added tests for this you should be able to validate that a custom widget that return an error schema and id get those errors shown.

**Review by heath-freenome — CHANGES_REQUESTED** [review]

File: packages/core/test/ObjectField.test.jsx [hunk 4]
```diff
@@ -5,8 +5,25 @@ import sinon from 'sinon';
 import { UI_GLOBAL_OPTIONS_KEY } from '@rjsf/utils';
 
 import SchemaField from '../src/components/fields/SchemaField';
+import ObjectField from '../src/components/fields/ObjectField';
 import { createFormComponent, createSandbox, submitForm } from './test_utils';
 
+const ObjectFieldTest = (props) => {
```
Comment (heath-freenome) [inline packages/core/test/ObjectField.test.jsx:11]: Thanks for adding these... Can you also test with a custom Widget inside an object?

File: packages/core/test/StringField.test.jsx [hunk 5]
```diff
@@ -3,9 +3,25 @@ import { Simulate } from 'react-dom/test-utils';
 import { fireEvent, act } from '@testing-library/react';
 import sinon from 'sinon';
 import { parseDateString, toDateString, TranslatableString, utcToLocal } from '@rjsf/utils';
+import StringField from '../src/components/fields/StringField';
 
 import { createFormComponent, createSandbox, getSelectedOptionValue, submitForm } from './test_utils';
 
+const StringFieldTest = (props) => {
```
Comment (heath-freenome) [inline packages/core/test/StringField.test.jsx:10]: Thanks for adding these... Can you also test with a custom Widget inside an string field?

**Review by heath-freenome — COMMENTED** [review]
Almost there, just a couple of small things. Also, can you resolve the conflicts?

File: packages/core/src/components/Form.tsx [hunk 6]
```diff
@@ -40,6 +40,7 @@ import _pick from 'lodash/pick';
 import _toPath from 'lodash/toPath';
 
 import getDefaultRegistry from '../getDefaultRegistry';
+import { forEach } from 'lodash';
```
Comment (heath-freenome) [inline packages/core/src/components/Form.tsx:43]: Oh shoot, sorry I missed this. Can you update the imports to explicitly import `forEach` like the ones above, putting it before `_get`

File: packages/core/src/components/Form.tsx [hunk 7]
```diff
@@ -38,6 +38,7 @@ import _get from 'lodash/get';
 import _isEmpty from 'lodash/isEmpty';
 import _pick from 'lodash/pick';
 import _toPath from 'lodash/toPath';
+import _forEach from 'lodash/forEach';
```
Comment (heath-freenome) [inline packages/core/src/components/Form.tsx:41]:
```suggestion
import _forEach from 'lodash/forEach';
import _isEmpty from 'lodash/isEmpty';
import _pick from 'lodash/pick';
import _toPath from 'lodash/toPath';
```

File: packages/core/src/components/Form.tsx [hunk 8]
```diff
@@ -35,6 +35,7 @@ import {
   Experimental_DefaultFormStateBehavior,
 } from '@rjsf/utils';
 import _get from 'lodash/get';
+import _forEach from 'lodash/forEach';
```
Comment (heath-freenome) [inline packages/core/src/components/Form.tsx:38]:
```suggestion
import _forEach from 'lodash/forEach';
import _get from 'lodash/get';
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -16,6 +16,12 @@ should change the heading of the (upcoming) version to include a major version b
 
 -->
 
+# 5.20.0
+
+## @rjsf/core 
+
+- Support allowing raising errors from within a custom whatever [#2718](https://github.com/rjsf-team/react-jsonschema-form/issues/2718)
+
```
Comment (heath-freenome) [inline CHANGELOG.md:24]:
```suggestion
## @rjsf/core 

- Support allowing raising errors from within a custom Widget [#2718](https://github.com/rjsf-team/react-jsonschema-form/issues/2718)

## @rjsf/utils

- Updated the `WidgetProps` type to add `es?: ErrorSchema<T>, id?: string` to the params of the `onChange` handler function

## Dev / docs / playground

-  Update the `custom-widget-fields.md` to add documentation for how to raise errors from a custom widget or field
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -16,6 +16,19 @@ should change the heading of the (upcoming) version to include a major version b
...
+- Support allowing raising errors from within a custom Widget [#2718](https://github.com/rjsf-team/react-jsonschema-form/issues/2718)
+
+## @rjsf/utils
+
+- Updated the `WidgetProps` type to add `es?: ErrorSchema<T>, id?: string` to the params of the `onChange` handler function
+
+## Dev / docs / playground
+
+-  Update the `custom-widget-fields.md` to add documentation for how to raise errors from a custom widget or field
```
Comment (heath-freenome) [inline CHANGELOG.md:31]:
```suggestion
-  Update the `custom-widget-fields.md` to add documentation for how to raise errors from a custom widget or field

```

**Review by heath-freenome — APPROVED** [review]

**Pull request merged**